File: ruby/lib/SpecificDamage.rb
# coding: utf-8
require_relative 'SpecificDamageToUI'
require_relative 'WeaponType'
require_relative 'Damage'

module Deepspace
  class SpecificDamage < Damage


    def initialize(nshields, wl)
      super(-1, nshields, wl)
    end

    def getUIversion
      SpecificDamageToUI.new(self)
    end
    
    def self.newCopy
      new(nShields, weapons)
    end
        
    def adjust(wl, sl)
      result = Array.new
      toRemove = Array.new(weapons)

      wl.each{|weapon|
        if toRemove.include?(weapon.type)
          result.push(weapon.type)
          toRemove.delete_at(toRemove.index(weapon.type))
        end
      }

      SpecificDamage.new(result, super.adjustShields(sl)) 
    end
        
    def arrayContainsType(w, t)
      index=-1
      for i in (0...w.length) do
        if(t == w[i].type)
          index = i
          break
        end
      end
      return index
    end    

    def discardWeapon(w)
      #Si borramos pero no está nos devuelve nil
      @weapons.delete(w.type)
    end
    
    def hasNoEffect
      nShields == 0 && weapons.length == 0
    end

    def to_s
      "Weapons #{weapons}; Shields: #{nShields};"
    end
    
    public :adjust
    private :arrayContainsType
    public_class_method :new
  end
end

File: java/src/deepspace/SpecificDamage.java
/*
 * Here comes the text of your license
 * Each line should be prefixed with  * 
 */
package deepspace;
import java.util.ArrayList;
/**
 *
 * @author yabir
 */
public class SpecificDamage extends Damage{
    private ArrayList<WeaponType> weapons;
            
    SpecificDamage(ArrayList<WeaponType> wl, int s){
	super(s);
        weapons = wl;
    }
    
    SpecificDamage(SpecificDamage d){
        this(d.getWeapons(), d.getNShields());
    }
    
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t){
	boolean done = false;
	int position = -1;

	for (int i =0; i < w.size() && !done ; i++){
	    if (w.get(i).getType() == t){
		position = i;
		done = true;
	    }
	}

	return position;
	
    }
    
    @Override
    public SpecificDamage adjust(ArrayList<Weapon> wl,  ArrayList<ShieldBooster> sl){
    
        ArrayList<WeaponType> result = new ArrayList<WeaponType>();
        ArrayList<Weapon> toRemove = new ArrayList<>(wl);

        for (Weapon weapon: wl){
            if (arrayContainsType(toRemove, weapon.getType()) != -1){
                result.add(weapon.getType());
                toRemove.remove(weapon.getType());
            }
        }
            
        return new SpecificDamage(result, Math.min(getNShields(), sl.size()));

    }
    
    @Override
    public boolean hasNoEffect(){
	return super.hasNoEffect() && weapons.isEmpty();

    }
    
    @Override
    SpecificDamageToUI getUIversion(){
        return new SpecificDamageToUI(this);
    }
    
    @Override
    public void discardWeapon(Weapon w){
        weapons.remove(w.getType());
    }
    
    @Override 
    SpecificDamage copy(){
        return new SpecificDamage(this);
    }
    
    public String toString() {
        return "weapons = " + weapons + "\n" +super.toString();
    }

}
